test/jdk/javax/swing/plaf/basic/BasicTreeUI/TreeCellRendererLeakTest.java

```diff
@@ -0,0 +1,208 @@
+
+        jTabbedPane1.setSelectedIndex(1);
+
+        frame = new JFrame();
+        frame.getContentPane().add(jTabbedPane1, java.awt.BorderLayout.CENTER);
+
+        frame.pack();
+        frame.setLocationRelativeTo(null);
+        frame.setVisible(true);
+    }// </editor-fold>
+
+    public static void main(String args[]) throws Exception {
+        try {
+            testDone = new CountDownLatch(1);
+            SwingUtilities.invokeAndWait(() -> {
+                TreeCellRendererLeakTest tf = new TreeCellRendererLeakTest();
+            });
+            testDone.await();
+        } finally {
+            SwingUtilities.invokeAndWait(() -> {
+                if (frame != null) {
+                    frame.dispose();
+                }
+            });
+        }
+    }
+
+    // Periodically cause a nodeChanged() for one of the nodes
+    public void runChanges() {
+        long count = 0;
+        long time = System.currentTimeMillis();
+        long tm = System.currentTimeMillis();
+        while ((tm - time) < (15 * 1000)) {
+            final long currentCount = count;
+            try {
+                SwingUtilities.invokeAndWait(new Runnable( ) {
+                    public void run( ) {
+                        DefaultTreeModel model = (DefaultTreeModel) jTree1.getModel();
+                        TreeNode root = (TreeNode) model.getRoot();
+                        DefaultMutableTreeNode n = (DefaultMutableTreeNode) model.getChild(root, 0);
+                        n.setUserObject("runcount " + currentCount);
+                        model.nodeChanged(n);
+                    }
+                });
+                count++;
+                Thread.sleep(1000);
+                tm = System.currentTimeMillis();
+                System.out.println("time elapsed " + (tm - time)/1000 + " s");
+            } catch (InterruptedException ex) {
+                break;
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+        testDone.countDown();
+    }
+
+    // Print number of uncollected JLabels
+    public void runInfo() {
+        long time = System.currentTimeMillis();
+        long initialCnt = smCount;
+        while ((System.currentTimeMillis() - time) < (15 * 1000)) {
+            System.gc();
+            System.out.println("Live JLabels:" + smCount);
+            try {
+                Thread.sleep(1000);
+            } catch (InterruptedException ex) {
+                ex.printStackTrace();
+            }
+        }
+
+        System.out.println("\ncleanedup LabelCount " + getCleanedUpLabelCount());
+        if (getCleanedUpLabelCount() == 0) {
+            throw new RuntimeException("TreeCellRenderer component leaked");
+        }
+    }
+
+    private int getCleanedUpLabelCount() {
+        int count = 0;
+        for (WeakReference<JLabel> ref : weakRefArrLabel) {
+            if (ref.get() == null) {
+                count++;
+            }
+        }
```

Comment:
You have to iterate over `weakRefArrLabel` in a synchronized block. (No, declaring `weakRefArrLabel` as `volatile` is not enough.)
```suggestion
        synchronized (weakRefArrLabel) {
            for (WeakReference<JLabel> ref : weakRefArrLabel) {
                if (ref.get() == null) {
                    count++;
                }
            }
        }
```

Reply:
Alternatively, you can use Stream API:

```java
    private long getCleanedUpLabelCount() {
        synchronized (weakRefArrLabel) {
            return weakRefArrLabel.stream()
                                  .filter(r -> r.get() != null)
                                  .count();
        }
    }
```